File: src/erlang/ch01/1.2/l1/src/modulo_examples.erl
-module(modulo_examples).
-export([hashing_example/2, cyclic_day/2, modexp/3, main/0]).

%% Simple hash bucket mapping
hashing_example(Key, Size) ->
    Key rem Size.

%% Cyclic wrap-around for weekdays (0–6)
cyclic_day(Start, Shift) ->
    (Start + Shift) rem 7.

%% Modular exponentiation (square-and-multiply)
modexp(A, B, M) ->
    modexp(A rem M, B, M, 1).

modexp(_, 0, _, Result) -> Result;
modexp(Base, Exp, M, Result) when Exp band 1 =:= 1 ->
    modexp((Base * Base) rem M, Exp bsr 1, M, (Result * Base) rem M);
modexp(Base, Exp, M, Result) ->
    modexp((Base * Base) rem M, Exp bsr 1, M, Result).

main() ->
    io:format("Hashing: key=1234, size=10 -> ~p~n", [hashing_example(1234, 10)]),
    io:format("Cyclic: Saturday(5)+4 -> ~p~n", [cyclic_day(5, 4)]),
    io:format("modexp(7,128,13) = ~p~n", [modexp(7, 128, 13)]).

File: archived/v1/src/c/ch01/1.2/modulo_examples.c
#include <stdio.h>
#include <stdint.h>

/* Simple hash bucket mapping */
int hashing_example(int key, int size) {
    return key % size;
}

/* Cyclic wrap-around for weekdays (0–6) */
int cyclic_day(int start, int shift) {
    return (start + shift) % 7;
}

/* Modular exponentiation (square-and-multiply) */
int modexp(int a, int b, int m) {
    long long result = 1;
    long long base = a % m;
    int exp = b;

    while (exp > 0) {
        if (exp & 1) {
            result = (result * base) % m;
        }
        base = (base * base) % m;
        exp >>= 1;
    }
    return (int)result;
}

int main(void) {
    printf("Hashing: key=1234, size=10 -> %d\n", hashing_example(1234, 10));
    printf("Cyclic: Saturday(5)+4 -> %d\n", cyclic_day(5, 4));
    printf("modexp(7,128,13) = %d\n", modexp(7, 128, 13));
    return 0;
}
